carp.client.core/src/commonMain/kotlin/dk/cachet/carp/client/domain/StudyRuntime.kt

```diff
@@ -108,24 +117,51 @@ class StudyRuntime private constructor(
      *
      * @return True in case deployment succeeded; false in case device could not yet be deployed (e.g., awaiting registration of other devices).
      * @throws UnsupportedOperationException in case deployment failed since not all necessary plugins to execute the study are available.
+     * @throws IllegalStateException in case data requested in the deployment cannot be collected on this client.
      */
-    suspend fun tryDeployment( deploymentService: DeploymentService ): Boolean
+    suspend fun tryDeployment( deploymentService: DeploymentService, dataCollector: DataCollector ): Boolean
     {
         val deploymentStatus = deploymentService.getStudyDeploymentStatus( studyDeploymentId )
-        val clientDeviceStatus = deploymentStatus.getDeviceStatus( device )
 
-        return tryDeployment( deploymentService, clientDeviceStatus )
+        return tryDeployment( deploymentService, dataCollector, deploymentStatus )
     }
 
-    private suspend fun tryDeployment( deploymentService: DeploymentService, deviceStatus: DeviceDeploymentStatus ): Boolean
+    private suspend fun tryDeployment(
+        deploymentService: DeploymentService,
+        dataCollector: DataCollector,
+        deploymentStatus: StudyDeploymentStatus
+    ): Boolean
     {
         // Early out in case state indicates the device is not yet ready to deploy.
+        val deviceStatus = deploymentStatus.getDeviceStatus( device )
         if ( deviceStatus !is DeviceDeploymentStatus.NotDeployed || !deviceStatus.isReadyForDeployment ) return false
 
-        // TODO: Get deployment and throw exception in case there are missing plugins to perform the operations (e.g., measurements).
+        // Get deployment information.
         val deployment = deploymentService.getDeviceDeploymentFor( studyDeploymentId, device.roleName )
+        check( deployment.deviceDescriptor == device )
         deploymentInformation = deployment
 
+        // Verify whether data can be collected for all requested measures.
+        for ( deviceTasks in deployment.getTasksPerDevice() )
```

Comment:
Maybe you could wrap the for loop and its content into a function for easier readability?

Reply:
I think that is a personal preference. I tend to not extract things out if there is no clear possibility for reuse. https://whatheco.de/2010/12/07/function-hell/

I gave this specific case some more thought and can't find a suitable place to extract the logic to. Not in `DataCollector` as it would introduce an unnecessarily strong link to `MasterDeviceDeployment`. It lives in the realm between `DataCollector` and `MasterDeviceDeployment`, which essentially is `StudyRuntime`. As a public function in `StudyRuntime` I think it leaks the abstraction it is trying to hide. As a private function in `StudyRuntime` it would only get called from one location.

I do agree this function is on the long side, but prefer to refactor/extract as it grows more and when it might make more sense.